application/src/main/java/org/thingsboard/server/service/cf/ctx/state/GeofencingArgumentEntry.java

```
@@ -0,0 +1,104 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+@Data
+@Slf4j
+public class GeofencingArgumentEntry implements ArgumentEntry {
+
+    private Map<EntityId, GeofencingZoneState> zoneStates;
+
+    private boolean forceResetPrevious;
+
+    public GeofencingArgumentEntry() {
+    }
+
+    public GeofencingArgumentEntry(Map<EntityId, KvEntry> entityIdkvEntryMap) {
+        this.zoneStates = toZones(entityIdkvEntryMap);
+    }
+
+    @Override
+    public ArgumentEntryType getType() {
+        return ArgumentEntryType.GEOFENCING;
+    }
+
+    @Override
+    public Object getValue() {
+        return zoneStates;
+    }
+
+    @Override
+    public boolean updateEntry(ArgumentEntry entry) {
+        if (!(entry instanceof GeofencingArgumentEntry geofencingArgumentEntry)) {
+            throw new IllegalArgumentException("Unsupported argument entry type for geofencing argument entry: " + entry.getType());
+        }
+        boolean updated = false;
+        for (var zoneEntry : geofencingArgumentEntry.getZoneStates().entrySet()) {
+            if (updateZone(zoneEntry)) {
+                updated = true;
+            }
+        }
+        return updated;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return zoneStates == null || zoneStates.isEmpty();
+    }
+
+    @Override
+    public TbelCfArg toTbelCfArg() {
+        return new TbelCfTsGeofencingArg(zoneStates);
+    }
+
+    private Map<EntityId, GeofencingZoneState> toZones(Map<EntityId, KvEntry> entityIdKvEntryMap) {
+        return entityIdKvEntryMap.entrySet().stream().map(entry -> {
+            try {
+                if (entry.getValue().getJsonValue().isEmpty()) {
+                    return null;
+                }
+                return Map.entry(entry.getKey(), new GeofencingZoneState(entry.getKey(), entry.getValue()));
+            } catch (Exception e) {
+                log.error("Failed to parse geofencing zone perimeter for entity id: {}", entry.getKey(), e);
```

Should we log at ERROR level with full stacktrace? I guess we will spam the logs. And why we at all ignore the error? Maybe put to debug event and fail the calculation?

---

I’ve decided that if multiple zones can be resolved, the calculation shouldn’t fail just because one of the zone states can’t be created (e.g., due to an attribute typo).

As for the logging level, I agree that DEBUG is more appropriate than ERROR in this case. 


I'm considering argument as invalid only if all zones failed to create:

```
@Override
public boolean isEmpty() {
    return zoneStates == null || zoneStates.isEmpty();
}
```